arch/s390/dfltcc.c

```diff
@@ -0,0 +1,778 @@
+/* dfltcc.c - IBM Z DEFLATE CONVERSION CALL support. */
+
+/*
+   Use the following commands to build zlib-ng with DFLTCC support:
+
+        $ ./configure --dfltcc
+        $ make
+*/
+
+#include <ctype.h>
+#include <inttypes.h>
+#include <stddef.h>
+#include <stdio.h>
+#include <stdint.h>
+#include "zbuild.h"
+#include "zutil.h"
+#include "deflate.h"
+#include "inftrees.h"
+#include "inflate.h"
+#include "dfltcc.h"
+#include "dfltcc_deflate.h"
+#ifdef S390_DFLTCC_USDT
+#include <sys/sdt.h>
+#endif
+
+/*
+   Tuning parameters.
+ */
+#ifndef DFLTCC_LEVEL_MASK
+#define DFLTCC_LEVEL_MASK 0x2
+#endif
+#ifndef DFLTCC_BLOCK_SIZE
+#define DFLTCC_BLOCK_SIZE 1048576
+#endif
+#ifndef DFLTCC_FIRST_FHT_BLOCK_SIZE
+#define DFLTCC_FIRST_FHT_BLOCK_SIZE 4096
+#endif
+#ifndef DFLTCC_DHT_MIN_SAMPLE_SIZE
+#define DFLTCC_DHT_MIN_SAMPLE_SIZE 4096
+#endif
+#ifndef DFLTCC_RIBM
+#define DFLTCC_RIBM 0
+#endif
+
+/*
+   C wrapper for the DEFLATE CONVERSION CALL instruction.
+ */
+typedef enum {
+    DFLTCC_CC_OK = 0,
+    DFLTCC_CC_OP1_TOO_SHORT = 1,
+    DFLTCC_CC_OP2_TOO_SHORT = 2,
+    DFLTCC_CC_OP2_CORRUPT = 2,
+    DFLTCC_CC_AGAIN = 3,
+} dfltcc_cc;
+
+#define DFLTCC_QAF 0
+#define DFLTCC_GDHT 1
+#define DFLTCC_CMPR 2
+#define DFLTCC_XPND 4
+#define HBT_CIRCULAR (1 << 7)
+#define HB_BITS 15
+#define HB_SIZE (1 << HB_BITS)
+#define DFLTCC_FACILITY 151
+
+static inline dfltcc_cc dfltcc(int fn, void *param,
+                               unsigned char **op1, size_t *len1, const unsigned char **op2, size_t *len2, void *hist)
+{
+    unsigned char *t2 = op1 ? *op1 : NULL;
+    size_t t3 = len1 ? *len1 : 0;
+    const unsigned char *t4 = op2 ? *op2 : NULL;
+    size_t t5 = len2 ? *len2 : 0;
+    register int r0 __asm__("r0") = fn;
+    register void *r1 __asm__("r1") = param;
+    register unsigned char *r2 __asm__("r2") = t2;
+    register size_t r3 __asm__("r3") = t3;
+    register const unsigned char *r4 __asm__("r4") = t4;
+    register size_t r5 __asm__("r5") = t5;
+    int cc;
+
+    __asm__ volatile(
+#ifdef S390_DFLTCC_USDT
```

Comment:
What is this USDT? I don't think it gets explained anywhere.

Reply:
If it is for debugging, we already have defines for enabling verbose debugging where the code will spit out a lot of information. This might be a case where that could be used instead?

Reply:
Yes, this is for debugging (https://myaut.github.io/dtrace-stap-book/app/proc.html). One has to have sys/sdt.h present to make use of it. Since the inserted probe points cost almost nothing, maybe I should simply replace S390_DFLTCC_USDT with something like HAVE_SYS_SDT_H?

Reply:
It can stay like it is, I was just wondering whether it would print debugging messages similar to assert for example.